src/main/java/org/manifold/compiler/front/PrimitiveNodeVertex.java

```diff
@@ -105,32 +200,47 @@ public void elaborate() throws Exception {
       throw new TypeMismatchException(
           typeConstructor, portType);
     }
-    
-    // check that both the input type and output type are tuple typevalues
-    // whose entries are PortTypeValue, then extract the identifier-typevalue
+
+    // look for PortTypeValues in the function signature,
+    // then extract the identifier-typevalue
     // pairs and add them to the port type map
+    log.debug("extracting input port types");
     extractPortTypes(portType.getInputType(), portTypeMap);
+    log.debug("extracting output port types");
     extractPortTypes(portType.getOutputType(), portTypeMap);
-    
-    Map<String, TypeValue> attributesMap = new HashMap<>();
-    ExpressionVertex attributesVertex = attributesEdge.getSource();
-    attributesVertex.elaborate();
-    if (!(attributesVertex.getType()
-        .isSubtypeOf(TypeTypeValue.getInstance()))) {
-      throw new TypeMismatchException(
-          TypeTypeValue.getInstance(),
-          attributesVertex.getType());
+
+    // look for non-PortTypeValues in the input type only,
+    // then extract the pairs and add to the attributes map
+    // TODO check for and disallow attributes in the output type
+    log.debug("extracting attributes");
+    extractAttributes(portType.getInputType(), attributesMap);
+
+    // TODO check for ports/attributes with exactly the same key
+
+    this.node = new NodeTypeValue(attributesMap, portTypeMap);
+    log.debug("constructed node type " + debugNodeType(node));
```

Comment:
We should probably be a little thriftier with our use of logging. Really noisy logs are about as good as no logs.